contracts/random/RandomValues.sol

```diff
@@ -0,0 +1,127 @@
+ */
+abstract contract RandomValues {
+    // Address of random seed provider contract.
+    // This value "immutable", and hence patched directly into bytecode when it is used.
+    // There will only ever be one random seed provider per chain. Hence, this value
+    // does not need to be changed.
+    RandomSeedProvider public immutable randomSeedProvider;
+
+    // Map of request id to fulfilment id.
+    mapping(uint256 requestId => uint256 fulfilmentId) private randCreationRequests;
+    // Map of request id to random source. Retaining the source allows for upgrade
+    // of sources inside the random seed provider contract.
+    mapping(uint256 requestId => address randomSource) private randCreationRequestsSource;
+
+    // Each request has a unique request id. The id is the current value
+    // of nextNonce. nextNonce is incremented for each request.
+    uint256 private nextNonce;
+
+    /**
+     * @notice Set the address of the random seed provider.
+     * @param _randomSeedProvider Address of random seed provider.
+     */
+    constructor(address _randomSeedProvider) {
+        randomSeedProvider = RandomSeedProvider(_randomSeedProvider);
+    }
+
+    /**
+     * @notice Register a request to generate a random value. This function should be called
+     *      when a game player has purchased an item the value of which is based on a random value.
+     * @return _randomRequestId A value that needs to be presented when fetching the random
+     *      value with fetchRandom.
+     */
+    function _requestRandomValueCreation() internal returns (uint256 _randomRequestId) {
+        uint256 randomFulfillmentIndex;
+        address randomSource;
+        (randomFulfillmentIndex, randomSource) = randomSeedProvider.requestRandomSeed();
+        _randomRequestId = nextNonce++;
+        randCreationRequests[_randomRequestId] = randomFulfillmentIndex;
+        randCreationRequestsSource[_randomRequestId] = randomSource;
+    }
+
+    /**
+     * @notice Fetch a random value that was requested using _requestRandomValueCreation.
+     * @dev The value is customised to this game, the game player, and the request by the game player.
+     *      This level of personalisation ensures that no two players end up with the same random value
+     *      and no game player will have the same random value twice.
+     * @param _randomRequestId The value returned by _requestRandomValueCreation.
+     * @return _randomValue The random number that the game can use.
+     */
+    function _fetchRandom(uint256 _randomRequestId) internal returns (bytes32 _randomValue) {
+        // Don't return the personalised seed directly. Otherwise there is a risk that
+        // the seed will be revealed, which would then compromise the security of calls
+        // to fetchRandomValues.
+        bytes32 seed = _fetchPersonalisedSeed(_randomRequestId);
+        _randomValue = keccak256(abi.encodePacked(seed, uint256(0)));
```

Comment:
Why pack with uint256(0)?

Reply:
The thinking behind this is that calling _fetchRandomValues returns 
keccak256(abi.encodePacked(seed, uint256(1)));
to 
keccak256(abi.encodePacked(seed, uint256(i)));

If _fetchRandomValue simply returned the output of _fetchPersonalisedSeed(_randomRequestId), there is a risk that a game studio would mistakenly:
- Call _fetchRandomValue
- The separately call _fetchRandomValues
Returning the pre-image of a set of random values. It would allow someone to simply generate the output of _fetchRandomValues. 
I realise that given the platform is a blockchain, once the seed from the RandomSeedProvider is locked in, anyone can guess the values returned by _fetchRandomValue and _fetchRandomValues. However, this makes it harder.

Reply:
My concern is that its very easy to simulate the tx before submitting it, while viewing the trace, which seems like it would negate any benefit here

Reply:
I guess you are right. It would provide no substantive benefit to anything but the most trivial analysis. The number returned by _fetchRandomValue and the values returned by _fetchRandomValues will all be different, which is the main thing. That is, the game can't mistakenly call _fetchRandomValue, and then separately _fetchRandomValues, and have any numbers be the same.

Reply:
I will remove the extra hashing.